zerver/webhooks/facebook/view.py

```diff
@@ -0,0 +1,114 @@
+        message = message + '\n* {verb}'.format(
+            verb=payload['entry'][0]['changes'][0]['value']['verb']
+        )
+        for id in payload['entry'][0]['changes'][0]['value']['target_ids']:
+            message = message + '\n  * {id}'.format(id=id)
+    return message
+
+def application_object(payload: Dict[Text, Any]) -> Text:
+    field = payload['entry'][0]['changes'][0]['field']
+    message = '**{field}** received'.format(field=field)
+    if field == 'plugin_comment':
+        message = message + '\n**{msg_user}:**\n```quote\n{message}\n```'.format(
+            msg_user=payload['entry'][0]['changes'][0]['value']['from']['name'],
+            message=payload['entry'][0]['changes'][0]['value']['message']
+        )
+    if field == 'plugin_comment_reply':
+        message = message + '\n**{prt_msg_user}:** (Parent)\n'\
+            '```quote\n{prt_message}\n```'.format(
+                prt_msg_user=payload['entry'][0]['changes'][0]['value']['parent']['from']['name'],
+                prt_message=payload['entry'][0]['changes'][0]['value']['parent']['message']
+            )
+        message = message + '\n**{cld_msg_user}:**\n```quote\n'\
+            '```quote\n{cld_message}\n```\n```'.format(
+                cld_msg_user=payload['entry'][0]['changes'][0]['value']['from']['name'],
+                cld_message=payload['entry'][0]['changes'][0]['value']['message']
+            )
+    return message
+
+@api_key_only_webhook_view("Facebook")
+@has_request_variables
+def api_facebook_webhook(request: HttpRequest, user_profile: UserProfile,
+                         stream: Text=REQ(default='Facebook'), token: Text=REQ()) -> HttpResponse:
+
+    if request.method == 'GET':  # facebook webhook verify
+        if request.GET.get("hub.mode") == 'subscribe':
+            if request.GET.get('hub.verify_token') == token:
+                return HttpResponse(request.GET.get('hub.challenge'))
+            else:
+                return json_error(('Error: Token is wrong'))  # nocoverage
+        else:
+            return json_error(('Error: Unsupported method'))  # nocoverage
+    else:  # facebook webhook receiver
+        payload = json.loads(request.body.decode("UTF-8"))
+        object = get_object(payload)
+        if object is not None:
+            body = get_body_based_on_object(object)(payload)
+            subject = object + " notification"
+            check_send_stream_message(user_profile, request.client,
+                                      stream, subject, body)
+        return json_success()
+
+OBJECTS_FUNCTION_MAPPER = {
+    'user': user_object,
+    'page': page_object,
+    'permissions': permissions_object,
+    'application': application_object
+}
+
+def get_object(payload: Dict[Text, Any]) -> Optional[Text]:
+    object = payload['object']
+    if object in OBJECTS_FUNCTION_MAPPER:
+        return object
+    raise UnknownObjectType(u"Object '{}' is unknown and cannot be handled".format(object))  # nocoverage
```

Comment:
Same advice as above:
1. Use `event` instead of `object` and `get_event` instead of `get_object`.
2. Any particular reason, why we aren't testing for that last line?

Reply:
2.
Maybe, this case return 500 error.
Is there a function to check 500 errors?

This integration was written with reference to [zerver/webhooks/gci](https://github.com/zulip/zulip/tree/master/zerver/webhooks/gci), but it looks like use # nocoverage...
Do I also have to fix this?

Reply:
@nyan-salmon: No worries, don't worry about it! I just ran a quick `git grep` and it turns out that last line isn't tested anywhere anyway, so it is not a huge deal! :)